infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,195 @@
+package org.corfudb.infrastructure.logreplication;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationEvent;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationState;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationStateType;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.IllegalTransitionException;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeEvent;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeState;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeStateType;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.sink.LogReplicationSinkEvent;
+import org.corfudb.runtime.LogReplication.LogReplicationSession;
+
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * This class manages the tasks submitted by the runtime FSM and the replication FSM.
+ * Both the FSMs have their own instances of this class to which tasks are submitted and executed.
+ *
+ * The threads are shared by all the sessions traversing an FSM.
+ */
+@Slf4j
+public class FsmTaskManager {
+
+    private ExecutorService fsmWorker;
+
+    private Map<LogReplicationSession, LinkedList<UUID>> sessionToRuntimeEventIdMap;
+
+    private Map<LogReplicationSession, LinkedList<UUID>> sessionToReplicationEventIdMap;
+
+    private Map<LogReplicationSession, LinkedList<UUID>> sessionToSinkEventIdMap;
+
+    public FsmTaskManager(String threadName, int threadCount) {
+        fsmWorker = Executors.newFixedThreadPool(threadCount, new ThreadFactoryBuilder().setNameFormat(threadName+"-%d").build());
+        sessionToRuntimeEventIdMap = new ConcurrentHashMap<>();
+        sessionToReplicationEventIdMap = new ConcurrentHashMap<>();
+        sessionToSinkEventIdMap = new ConcurrentHashMap<>();
+    }
+
+    public <E> void addTask(E event, FsmEventType fsm) {
```

Comment:
Why is a separate event type parameter required?  We can use check the type using instanceof

Reply:
We'll need to use reflection to check the type...which is not so efficient.